FrameMetadata/FrameMetadata/ViewController.m

```diff
@@ -0,0 +1,336 @@
+//
+//  ViewController.m
+//  Lets-Build-Publisher
+//
+//  Copyright (c) 2013 TokBox, Inc. All rights reserved.
+//
+
+#import "ViewController.h"
+#import <OpenTok/OpenTok.h>
+#import "TBExampleVideoCapture.h"
+#import "TBExampleVideoRender.h"
+
+@interface ViewController ()
+<OTSessionDelegate, OTSubscriberKitDelegate, OTPublisherDelegate, TBRendererDelegate, TBExampleVideoCaptureDelegate>
+
+@end
+
+@implementation ViewController {
+    OTSession* _session;
+    OTPublisher* _publisher;
+    OTSubscriber* _subscriber;
+    
+    TBExampleVideoCapture* _defaultVideoCapture;
+    
+    TBExampleVideoRender* _subscriberVideoRenderView;
+    TBExampleVideoRender* _publisherVideoRenderView;
+    
+    UILabel *metadataLabel;
+}
+static double widgetHeight = 240;
+static double widgetWidth = 320;
+
+// *** Fill the following variables using your own Project info  ***
+// ***          https://dashboard.tokbox.com/projects            ***
+// Replace with your OpenTok API key
+static NSString* const kApiKey = @"";
+// Replace with your generated session ID
+static NSString* const kSessionId = @"";
+// Replace with your generated token
+static NSString* const kToken = @"";
+
+#pragma mark - View lifecycle
+
+- (void)viewDidLoad
+{
+    [super viewDidLoad];
+    
+    // Step 1: As the view comes into the foreground, initialize a new instance
+    // of OTSession and begin the connection process.
+    _session = [[OTSession alloc] initWithApiKey:kApiKey
+                                       sessionId:kSessionId
+                                           delegate:self];
+    
+    metadataLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 100, 30)];
+    [metadataLabel setTextColor:[UIColor whiteColor]];
+    [metadataLabel setFont:[UIFont systemFontOfSize:12]];
+    
+    [self doConnect];
+}
+
+- (BOOL)prefersStatusBarHidden
+{
+    return YES;
+}
+
+- (BOOL)shouldAutorotateToInterfaceOrientation:
+(UIInterfaceOrientation)interfaceOrientation
+{
+    return UIUserInterfaceIdiomPhone != [[UIDevice currentDevice] userInterfaceIdiom];
+}
+#pragma mark - OpenTok methods
+
+/**
+ * Asynchronously begins the session connect process. Some time later, we will
+ * expect a delegate method to call us back with the results of this action.
+ */
+- (void)doConnect
+{
+    OTError *error = nil;
+    [_session connectWithToken:kToken error:&error];
+    if (error)
+    {
+        [self showAlert:[error localizedDescription]];
+    }
+}
+
+/**
+ * Sets up an instance of OTPublisher to use with this session. OTPubilsher
+ * binds to the device camera and microphone, and will provide A/V streams
+ * to the OpenTok session.
+ */
+- (void)doPublish
+{
+    OTPublisherSettings *pubSettings = [[OTPublisherSettings alloc] init];
+    pubSettings.name = [[UIDevice currentDevice] name];
+    _publisher = [[OTPublisher alloc]
+                  initWithDelegate:self settings:pubSettings];
+    
+    TBExampleVideoCapture* videoCapture =
```

Comment:
once the new delgate approach is added, please rename the name of the class to something more relevant like `TBFrameMetadataCapturer`

Reply:
Looks like we can't do it this way. If we have `TBFrameMetadataCapturer` inherited from the `TBExampleVideoRender`. Eventually, we still need to call `[self.delegate finishPreparingFrame:]` inside `didOutputSampleBuffer` so that `TBExampleVideoRender` could not remain unchanged. If we put @protocol and @property to `TBFrameMetadataCapturer`, the `TBExampleVideoRender` could not access the delegate object which sits at the child class. Being said, directly modifying the `TBExampleVideoRender` is the only way now as far as I see. The whole point is: `TBExampleVideoRender` could not remain unchanged.

Reply:
But I think we weren't going to use inheritance. We need to modify TBExampleCapturer (from the other sample) to add delegate support.
Then, we just need to add a delegate in this sample that will add the metadata. Sorry If I confused you talking about `TBFrameMetadataCapturer`, in fact, it should be something like `TBFrameCapturerMetadataDelegate`. That class will be a delegate of `TBExampleCapturer` and will provide the frame metadata